EventBus/EventBus.swift

```diff
@@ -8,133 +8,413 @@
+    func eventBus<T>(_ eventBus: EventBus, receivedEvent: T.Type)
+}
+
+internal struct DefaultLogHandler: LogHandler {
+    func eventBus<T>(_ eventBus: EventBus, receivedEvent eventType: T.Type) {
+        #if DEBUG
+            print("\(eventBus): Received event '\(eventType)'.")
+        #endif
+    }
+}
+
+/// A type-safe event bus.
 public class EventBus {
 
-    struct WeakBox : Hashable {
-        weak var inner: AnyObject?
+    internal struct WeakBox: Hashable {
+        internal weak var inner: AnyObject?
 
-        init(_ inner: AnyObject) {
+        internal init(_ inner: AnyObject) {
             self.inner = inner
         }
 
-        static func == (lhs: WeakBox, rhs: WeakBox) -> Bool {
+        internal static func == (lhs: WeakBox, rhs: WeakBox) -> Bool {
             return lhs.inner === rhs.inner
         }
 
-        var hashValue: Int {
+        internal var hashValue: Int {
             guard let inner = self.inner else {
                 return 0
             }
             return ObjectIdentifier(inner).hashValue
         }
     }
 
+    /// A global shared event bus configured using default options.
     public static let shared: EventBus = EventBus()
 
-    fileprivate var registered: [ObjectIdentifier: String]?
+    /// The event bus' configuration options.
+    public let options: Options
+
+    internal var errorHandler: ErrorHandler = DefaultErrorHandler()
+    internal var logHandler: LogHandler = DefaultLogHandler()
+
+    fileprivate var registered: [ObjectIdentifier: Any] = [:]
     fileprivate var subscribed: [ObjectIdentifier: Set<WeakBox>] = [:]
     fileprivate var chained: Set<WeakBox> = []
 
     fileprivate let serialQueue: DispatchQueue = DispatchQueue(label: "com.regexident.eventbus")
     fileprivate let queue: DispatchQueue
 
-    public init(strict: Bool = false, queue: DispatchQueue = .global()) {
-        self.registered = strict ? [:] : nil
+    /// Creates an event bus with a given configuration and dispatch queue.
+    ///
+    /// - Parameters:
+    ///   - options: the event bus' options
+    ///   - queue: the dispatch queue to notify subscribers on
+    public init(options: Options? = nil, queue: DispatchQueue = .global()) {
         self.queue = queue
+        self.options = options ?? Options()
     }
 
-    fileprivate func validateSubscriber<T>(subscriber: T) {
+    /// The event types the event bus is registered for.
+    public var registeredEventTypes: [Any] {
+        return Array(self.registered.values)
+    }
+
+    fileprivate func warnIfNonClass<T>(_ subscriber: T) {
         // Related bug: https://bugs.swift.org/browse/SR-4420:
-        if !(type(of: subscriber as Any) is AnyClass) {
-            let message = "Expected class, found struct/enum: \(subscriber)"
-            #if DEBUG
-                fatalError(message)
-            #else
-                print("Warning:", message)
-            #endif
+        guard !(type(of: subscriber as Any) is AnyClass) else {
```

Comment:
According to https://bugs.swift.org/browse/SR-4420 it has been fixed?

Reply:
Yes. This (the highlighted line) used to not work for a version of Swift, or two, as seen in the bug report of mine.

The run-time check (for which the semantics have been fixed again, by now) is necessary nevertheless, though, as changing …

```swift
func add<T>(subscriber: T, for eventType: T.Type)
```

… to …

```swift
func add<T: AnyObject>(subscriber: T, for eventType: T.Type)
```

… makes `T` be interpreted as a concrete type, instead of a protocol object.